### PR TITLE
fix(variable): change 'value' type to Dynamic

### PR DESCRIPTION
**Pull request by mitchnielsen**
## Summary

Changes the Variable 'value' type from string to Dynamic.

In https://github.com/PrefectHQ/prefect/pull/13543 and associated changes around May 2024, Variables were updated from simple strings to JSON objects. The Terraform provider has still been treating them as strings, so when folks tried to put JSON-compatible values in them, Terraform would fail to work with them as found in https://github.com/PrefectHQ/terraform-provider-prefect/issues/254

That said, this doesn't mean you can always expect a full JSON object in response. It could be individual JSON-compatible types, like `bool` or `int`. So we actually need a Dynamic attribute type here.

Related to https://linear.app/prefect/issue/PLA-247/changing-variable-to-a-json-value-in-the-ui-makes-next-terraform-run

Related to https://github.com/PrefectHQ/terraform-provider-prefect/issues/254

## References

- https://developer.hashicorp.com/terraform/language/expressions/types
- https://developer.hashicorp.com/terraform/language/expressions/type-constraints
- https://developer.hashicorp.com/terraform/plugin/framework/handling-data/dynamic-data
- https://developer.hashicorp.com/terraform/plugin/framework/handling-data/attributes/dynamic

From https://docs.prefect.io/3.0/develop/variables:

> (Variables) are named, mutable values of any JSON type

Also, you can see the accepted types for the `value` field here:

```
$ curl -s https://api.prefect.cloud/api/openapi.json | jq '.components.schemas.Variable.properties.value.anyOf'
$ curl -s https://api.prefect.cloud/api/openapi.json | jq '.components.schemas.Variable.properties.value.anyOf[]'
{
  "type": "string",
  "maxLength": 5000
}
{
  "type": "integer"
}
{
  "type": "number"
}
{
  "type": "boolean"
}
{
  "type": "object"
}
{
  "items": {},
  "type": "array"
}
```

### CODE REVIEW

**Comment by mitchnielsen** [comment]
So changing this to a JSON type may not be the full fix here. As the API response shows, the value here can be almost anything: string, integer, number, boolean, object, and array. So if I put `5` in the Variable, the provider will fail to convert that into `map[string]interface`. Need to think about this some more.

**Comment by mitchnielsen** [comment]
We may need to use a dynamic attribute here: https://developer.hashicorp.com/terraform/plugin/framework/handling-data/attributes/dynamic

File: docs/resources/variable.md [hunk 1]
```diff
@@ -25,7 +25,7 @@ resource "prefect_variable" "example" {
 ### Required
 
 - `name` (String) Name of the variable
-- `value` (String) Value of the variable
+- `value` (Dynamic) Value of the variable, supported Terraform value types: string, number, bool, tuple, object
```
Comment (parkedwards) [inline docs/resources/variable.md:28]: brutal!

File: internal/provider/datasources/variable.go [hunk 2]
```diff
@@ -182,3 +194,60 @@ func (d *VariableDataSource) Read(ctx context.Context, req datasource.ReadReques
 		return
 	}
 }
+
+// getDynamicValue converts the 'value' attribute from a native Go type to a DynamicValue.
+func getDynamicValue(model VariableDataSourceModel, variable *api.Variable) (basetypes.DynamicValue, diag.Diagnostics) {
```
Comment (parkedwards) [inline internal/provider/datasources/variable.go:199]: v nice

**Review by parkedwards — APPROVED** [review]

**Pull request merged**